crates/apollo-compiler/src/diagnostic.rs

```diff
@@ -300,24 +300,25 @@ impl ariadne::Cache<FileId> for Cache<'_> {
         }
     }
 
-    fn display<'a>(&self, file_id: &'a FileId) -> Option<Box<dyn fmt::Display + 'a>> {
-        if *file_id != FileId::NONE {
-            struct Path(Arc<SourceFile>);
-            impl fmt::Display for Path {
-                fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-                    self.0.path().display().fmt(f)
+    fn display<'a>(&self, file_id: &'a FileId) -> Option<impl fmt::Display + 'a> {
+        enum Path {
+            SourceFile(Arc<SourceFile>),
+            NoSourceFile,
+        }
+        impl fmt::Display for Path {
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                match self {
+                    Path::SourceFile(source_file) => source_file.path().display().fmt(f),
+                    Path::NoSourceFile => f.write_str("(no source file)"),
                 }
             }
+        }
+
+        if *file_id != FileId::NONE {
             let source_file = self.0.get(file_id)?;
-            Some(Box::new(Path(source_file.clone())))
+            Some(Box::new(Path::SourceFile(source_file.clone())))
```

Comment:
```suggestion
            Some(Path::SourceFile(source_file.clone()))
```
we probably don't need to box this anymore which is nice

Reply:
just spotted that this morning too. removed the Box in fn fetch as well 🙏